File: twpqm.c
/*
 * Three Way Partition Quick median take an array of n float 
 * and returns the median value. It combines the three way partition 
 * algorithm or "Dutch Flag" from Edsger Dijkstra and the Quick Select 
 * from Tony Hoare. Strategy is divide and Conquer.
 * The array is split iteratively in 3 parts: < = > of pivot 
 * Outperforms QuickSelect algorithm for arrays with duplicates or pre sorted.
 * Warning: in place value shuffling
 * Written by Emmanuel Brandt 2019-04
 * @param a array of float
 * @param n array size
 * @return median as float
 */
 
/* 
** Quickmedian takes an array a of n float and return the median value. 
** Use quicksort on half of the array, iterative version.
**/
float quickmedian1(float *a, int n) 
{	
	int i;
	int k = n / 2;		// size to sort
	int pindex;			// pivot index
	int left = 0; 		// left index
	int right = n - 1; 	// right index
	register float pivot, tmp;

	while (left < right) { //we stop when our indicies have crossed
		pindex = (left + right) / 2; // pivot selection, this can be whatever
		pivot = a[pindex];
		a[pindex] = a[right];
		a[right] = pivot; // SWAP(pivot,right)

		for (i = pindex = left; i < right; i++) {
			if (a[i] < pivot) { // SWAP
				tmp = a[pindex];
				a[pindex] = a[i];
				a[i] = tmp;
				pindex++;
			}
		}
		a[right] = a[pindex];
		a[pindex] = pivot; // SWAP

		if (pindex < k)
			left = pindex + 1;
		else // pindex >= k
			right = pindex;
	}
	return a[k];
}

File: twpqm.jl
"""
	twpqs(a::Array) -> median

Three-way partition Quick Median, in-place sorting.
A combination of Quick Median from Tony Hoare and
the Deutch Flag algorithm from Edsger Dijkstra.
- `a::Array`: even length.
"""
function twpqm(a::Array)
    n = length(a)
    iseven(n) && return

    left = l = j = 1
    right = r = n
    k = 1 + n >> 1 # final median index
    pivot = a[k] # initial pivot value

    # Partion elements: lower than, equal, greater than pivot
    @inbounds while left <= right
        while j <= r#
            if a[j] < pivot
                a[j], a[l] = a[l], a[j]
                l += 1
                j += 1
            elseif a[j] > pivot
                a[j], a[r] = a[r], a[j]
                r -= 1
            else
                j += 1
            end
        end

        if k > r
            left = l = j = r + 1
            r = right
        elseif k < l
            r = right = l - 1
            l = j = left
        else
            return pivot
        end

        pivot = a[(l+r)>>1]
    end
end

using Statistics, Combinatorics, BenchmarkTools

function ttwpqm()
	A = rand(Float32.(1:9),99)
	twpqm(A)
end

function t3wp(n::Int)
    i = 0
    for a in permutations(1:n)
    #for j = 1:n a = rand(1:9, 99)
        m1 = twpqm(a)
        m2 = round(Int, median(a))
        if m1 != m2
            println("$a twpqm $m1 median $m2")
            break
        end
        i += 1
    end
    println(i)
end
